README.md
Node Airbrake
=============

[![Circle CI](https://circleci.com/gh/airbrake/node-airbrake.svg?style=shield)](https://circleci.com/gh/airbrake/node-airbrake)
[![npm version](https://badge.fury.io/js/airbrake.svg)](https://badge.fury.io/js/airbrake)
[![Documentation Status](http://inch-ci.org/github/airbrake/node-airbrake.svg?branch=master)](http://inch-ci.org/github/airbrake/node-airbrake)
[![Downloads](https://img.shields.io/npm/dt/airbrake.svg)](https://www.npmjs.com/package/airbrake)

**DEPRECATION WARNING: Node Airbrake is deprecated in favour of @airbrake/node
(https://www.npmjs.com/package/@airbrake/node). Please migrate to it as soon as
possible. See [migration guide](https://airbrake.io/docs/performance-monitoring/updating-from-deprecated-libraries-for-node/).**

![Node Airbrake][arthur-node]

* [Node Airbrake README][node-airbrake]

Introduction
------------

**DEPRECATION WARNING: Node Airbrake is deprecated in favour of @airbrake/node
(https://www.npmjs.com/package/@airbrake/node). Please migrate to it as soon as
possible.**

_Node Airbrake_ is a Node.js notifier for [Airbrake][airbrake-io], the leading
exception reporting service. The library provides minimalist API that enables
the ability to send _any_ Node exception to the Airbrake dashboard. Node
Airbrake provides out-of-box integration with the Express web framework.

Key features
------------

![The Airbrake Dashboard][dashboard]

* Send chosen environment variables (whitelist or blacklist)
* Detect and fix circular references in error context information
* Support for all features of the [2.1 notification API][2.1api]
* Support for [long-stack-traces][]
* Optional auto-handler for `uncaughtException` events
* Provides notification URL linking to Airbrake in `notify()` callback
* Timeout Airbrake requests after 30 seconds, you never know
* Express web application framework support
* hapi web application framework support

Installation
------------

### NPM

Add the Node Airbrake package to your `package.json`:

```js
{
  "dependencies": {
    "airbrake": "^2.1.3"
  }
}
```

### Manual

Invoke the following command from your terminal:

```sh
npm install airbrake
```

Examples
--------

### Basic example

This is the minimal example that you can use to test Node Airbrake with your
project. The common use case for this module is to catch all `'uncaughtException'`
events on the `process` object and send them to Airbrake:

```js
var airbrake = require('airbrake').createClient(
  '105138', // Project ID
  'fd04e13d806a90f96614ad8e529b2822' // Project key
);
airbrake.handleExceptions();

throw new Error('I am an uncaught exception');
```

Note: the above will re-throw the exception after it has been successfully
delivered to Airbrake, causing your process to exit with status 1.

This can optionally be disabled by passing false to `handleExceptions` (not
recommended):

```js
airbrake.handleExceptions(false);
```

### Filtering errors

There may be some errors thrown in your application that you're not interested in sending to Airbrake, such as errors thrown by 3rd-party libraries.

The Airbrake notifier makes it simple to ignore this chaff while still processing legitimate errors. Add filters to the notifier by providing filter functions to `addFilter`.

`addFilter` accepts the entire [error notice](https://airbrake.io/docs/#create-notice-v3) to be sent to Airbrake, and provides access to the `context`, `environment`, `params`, and `session` values submitted with the notice, as well as the single-element `errors` array with its `backtrace` element and associated backtrace lines.

The return value of the filter function determines whether or not the error notice will be submitted.
  * If null value is returned, the notice is ignored.
  * Otherwise returned notice will be submitted.

An error notice must pass all provided filters to be submitted.

In the following example errors triggered with a message of 'this should not be posted to airbrake' will be ignored:

```js
airbrake.addFilter(function(notice) {
  if (notice.errors[0].message === 'this should not be posted to airbrake') {
    // Ignore errors with this messsage
    return null;
  }
  return notice;
});
```

Filters can be also used to modify notice payload, e.g. to set environment and application version:

```js
airbrake.addFilter(function(notice) {
  notice.context.environment = 'production';
  notice.context.version = '1.2.3';
  return notice;
});
```

### Manual error delivery

If you want more control over the delivery of your errors, you can also
manually submit errors to Airbrake:

```js
var airbrake = require('airbrake').createClient("your project ID", "your api key");
var err = new Error('Something went terribly wrong');
airbrake.notify(err, function(err, url) {
  if (err) throw err;

  // Error has been delivered, url links to the error in airbrake
});
```

### Severity

[Severity](https://airbrake.io/docs/airbrake-faq/what-is-severity/) allows
categorizing how severe an error is. By default, it's set to `error`. To
redefine severity, simply set an error's `severity` property. For example:

```js
var err = new Error();
err.severity = 'critical';

airbrake.notify(err);
```

### Express integration

The library provides out-of-box integration with the Express framework. It
supports even old Express versions (starting from `2.x`). Select your version
below and configure accordingly.

#### Express 4.x

```js
var airbrake = require('airbrake').createClient("your project ID", "your api key");
app.use(airbrake.expressHandler());
```

#### Express 3.x

```js
var airbrake = require('airbrake').createClient("your project ID", "your api key");
app.use(app.router);
app.use(airbrake.expressHandler());
```

#### Express 2.x

```js
var airbrake = require('airbrake').createClient("your project ID", "your api key");
app.error(airbrake.expressHandler());
```

### hapi integration

The library provides out-of-box integration with the hapi framework. To
integrate Airbrake with a hapi application simply install our handler:

```js
const Hapi = require('hapi');
const server = new Hapi.Server();
const Airbrake = require('airbrake').createClient(
  "your project ID",
  "your api key"
);

Airbrake.env = 'production';

server.register(Airbrake.hapiHandler(), err => {
  if (err) {
    throw err;
  }
});
```

API
---

## Adding context to errors

The `notify()` method automatically adds the following context information to
each delivered error:

* **error.class:** (`err.type` string if set, or `'Error'`)
* **error.message:** (`err.message` string)
* **error.backtrace:** (`err.stack` as parsed by [stack-trace][])
* **error.severity:** (`err.severity` defaults to `error`)
* **request.url:** (`err.url`, see `airbrake.url`);
* **request.component:** (`err.component` string if set);
* **request.action:** (`err.action` string if set);
* **request.cgi-data:** (`process.env`, merged all other properties of `err`)
* **request.params:** (`err.params` object if set)
* **request.session:** (`err.session` object if set)
* **server-environment.project-root:** (`airbrake.projectRoot` string if set)
* **server-environment.environment-name:** (`airbrake.env` string)
* **server-environment.app-version:** (`airbrake.appVersion string if set)

You can add additional context information by modifying the error properties
listed above:

``` javascript
var airbrake = require('airbrake').createClient("your project ID", "your api key");
var http = require('http');

http.createServer(function(req, res) {
  if (req.headers['X-Secret'] !== 'my secret') {
    var err = new Error('403 - Permission denied');
    req.writeHead(403);
    req.end(err.message);

    err.url = req.url;
    err.params = {ip: req.socket.remoteAddress};
    airbrake.notify(err);
  }
});
```

Unfortunately `uncaughtException` events cannot be traced back to particular
requests, so you should still try to handle errors where they occur.

[stack-trace]: https://github.com/felixge/node-stack-trace

## Tracking deployments

This client supports Airbrake's [deployment tracking][]:

``` javascript
var airbrake = require('airbrake').createClient("your project ID", "your api key");
var deployment = {
  rev: '98103a8fa850d5eaf3666e419d8a0a93e535b1b2',
  repo: 'git@github.com:felixge/node-airbrake.git',
};

airbrake.trackDeployment(deployment, function(err, params) {
  if (err) {
    throw err;
  }

  console.log('Tracked deployment of %s to %s', params.rev, params.env);
});
```

Check out the `airbrake.trackDeployment()` API docs below for a list of all
options.

Configuration
-------------

### var airbrake = Airbrake.createClient(projectId, key, [env])

`Airbrake.createClient()` returns a new Airbrake instance.

Options
* `projectId` - Your application's Airbrake project ID.
* `key` - Your application's Airbrake API key.
* `env` - The name of the server environment this is running in.

### airbrake.projectId = null

Your application's Airbrake project ID.

### airbrake.key = null

Your application's Airbrake API key.

### airbrake.env = process.env.NODE_ENV || 'development'

The name of the server environment this is running in.

### airbrake.host = 'https://' + os.hostname()

The base url for errors. If `err.url` is not set, `airbrake.host` is used
instead. If `err.url` is a relative url starting with `'/'`, it is appended
to `airbrake.host`. If `err.url` is an absolute url, `airbrake.host` is ignored.

### airbrake.projectRoot = process.cwd()

The root directory of this project.

### airbrake.appVersion = null

The version of this app. Set to a semantic version number, or leave unset.

### airbrake.protocol = 'https'

The protocol to use.

### airbrake.timeout = 30 * 1000

The timeout after which to give up trying to notify Airbrake in ms.

### airbrake.proxy = null

The HTTP/HTTPS proxy to use when making requests.

### airbrake.requestOptions = {}

Additional request options that are merged with the default set of options that are passed to `request` during `notify()` and `trackDeployment()`.

### airbrake.whiteListKeys = []

Names of environment variables to send.

### airbrake.blackListKeys = []

Names of environment variables to filter out.

### airbrake.handleExceptions()

Registers a `process.on('uncaughtException')` listener. When an uncaught
exception occurs, the error is sent to Airbrake, and then re-thrown to
kill the process.

### airbrake.handlePromiseRejections()

Registers a `process.on('unhandledRejection')` listener. When an uncaught
exception occurs inside a promise, the error is sent to Airbrake, and then re-thrown to
kill the process.

### airbrake.expressHandler(disableUncaughtException)

A custom error handler that is used with Express. Integrate with Express
middleware using `app.use()`.

Options:
* `disableUncaughtException`: Disables re-throwing and killing process on uncaught exception.

### airbrake.notify(err, [cb])

Sends the given `err` to airbrake.

The callback parameter receives two arguments, `err, url`. `err` is set if
the delivery to Airbrake failed.

If no `cb` is given, and the delivery fails, an `error` event is emitted. If
there is no listener for this event, node will kill the process as well. This
is done to avoid silent error delivery failure.

### airbrake.trackDeployment([params, [cb]])

Notifies Airbrake about a deployment. `params` is an object with the following
options:

* `env:` The environment being deployed, defaults to `airbrake.env`.
* `user:` The user doing the deployment, defaults to `process.env.USER`.
* `repo:` The github url of this repository. Defaults to `git config --get remote.origin.url`.
* `rev:` The revision of this deployment. Defaults to `git rev-parse HEAD`.

Additional notes
----------------

### Exception limit

The maximum size of an exception is 64KB. Exceptions that exceed this limit will
be truncated to fit the size.

License
-------

The library was originally created by [Felix Geisendörfer](https://github.com/felixge).
The project uses the MIT License. See LICENSE.md for details.

[arthur-node]: http://s3.amazonaws.com/airbrake-github-assets/node-airbrake/arthur-node.jpeg
[node-airbrake]: https://github.com/airbrake/node-airbrake
[airbrake-io]: https://airbrake.io
[long-stack-traces]: https://github.com/tlrobinson/long-stack-traces
[2.1api]: http://help.airbrake.io/kb/api-2/notifier-api-version-21
[screenshot]: https://github.com/airbrake/node-airbrake/raw/master/screenshot.png
[deployment tracking]: https://airbrake.io/docs/airbrake-faq/deploy-tracking/
[dashboard]: http://s3.amazonaws.com/airbrake-github-assets/node-airbrake/airbrake-dashboard.png


CHANGELOG.md
Node Airbrake Changelog
=======================

### master

### [v2.1.3][v2.1.3] (April 23, 2020)
* Updated migration guide link

### [v2.1.2][v2.1.2] (April 23, 2020)

* Updated deprecation notice

### [v2.1.1][v2.1.1] (September 27, 2017)

* NODE AIRBRAKE IS DEPRECATED IN FAVOUR OF AIRBRAKE JS
  (https://github.com/airbrake/airbrake-js)
  ([#152](https://github.com/airbrake/node-airbrake/pull/152))

### [v2.1.0][v2.1.0] (July 25, 2017)

* Added `airbrake.handlePromiseRejections()`
  ([#147](https://github.com/airbrake/node-airbrake/pull/147))

### [v2.0.1][v2.0.1] (May 16, 2017)

* Started sending error severity (defaults to `error`)
  ([#143](https://github.com/airbrake/node-airbrake/pull/143))

### [v2.0.0][v2.0.0] (April 17, 2017)

* **IMPORTANT:** dropped support for Node.js <= v0.12
* Dropped `sync-exec` dependency in favour of native `child_process.execSync`
  ([#137](https://github.com/airbrake/node-airbrake/pull/137))
* Dropped `json-stringify-safe` dependency in favour of native `JSON.stringify`
  ([#138](https://github.com/airbrake/node-airbrake/pull/138))

### [v1.3.0][v1.3.0] (April 11, 2017)

* Implemented payload truncation (notices over 64KB are not accepted by the
  Airbrake API and will be truncated to fit the
  limit) ([#134](https://github.com/airbrake/node-airbrake/pull/134))

### [v1.2.2][v1.2.2] (March 10, 2017)

* Updated `request` dependency to `^2.81.0`
  ([#129](https://github.com/airbrake/node-airbrake/pull/129))

### [v1.2.1][v1.2.1] (February 15, 2017)

* Fixed bug when client crashes when tries to notify due to missing `projectId`
  or `projectKey`. These values are mandatory to create a new client now
  ([#125](https://github.com/airbrake/node-airbrake/pull/125))

### [v1.2.0][v1.2.0] (Semptember 8, 2016)

* **IMPORTANT:** Removed the `developmentEnvironments` config option
  ([#119](https://github.com/airbrake/node-airbrake/pull/119))
* Express & hapi integrations send `component` and `action` info now
  ([#120](https://github.com/airbrake/node-airbrake/pull/120))

### [v1.1.0][v1.1.0] (July 30, 2016)

* **IMPORTANT:** Added support for the Hapi framework
  ([#112](https://github.com/airbrake/node-airbrake/pull/112))

### [v1.0.3][v1.0.3] (June 24, 2016)

* Stopped depending on `lodash` and started depending on `lodash.merge`
  ([#102](https://github.com/airbrake/node-airbrake/pull/102))

### [v1.0.2][v1.0.2] (May 26, 2016)

* Change notifier name from `airbrake` to `node-airbrake`
  ([#96](https://github.com/airbrake/node-airbrake/pull/96))

### [v1.0.1][v1.0.1] (April 5, 2016)

* replace hasish with lodash ([#89][89])
* automatically parse repo and rev from local repo ([#88][88])
* stop defaulting to empty string when serializing line numbers

### [v1.0.0][v1.0.0] (March 2, 2016)

* update to JSON API ([#86][86])

### [v0.4.1][v0.4.1] (February 23, 2016)
### [v0.4.0][v0.4.0] (February 23, 2016)

* start sending user agent info from Express handler ([#79][79])
* make error handler's killing of process customizable ([#45][45])
* return a Buffer object from `notifyXML` to fix Errbit incompatibility ([#51][51])
* start safely serializing values with `json-stringify-safe` ([#60][60])
* remove hostname from the docs ([#61][61])
* update test suite to use Express 4.x ([#76][76])
* start using Circle CI ([#78][78])
* allow items to be excluded from CGI data ([#65][65])
* documentation fixes ([#68][68] and [#69][69])
* add ability to ignore exceptions ([#74][74])
* update default protocol to https ([#71][71])
* remove uncaught exception from tests ([#72][72])
* refactor `notify` into `_sendRequest` ([#73][73])
* make filtering of environment variables customizable ([#84][84])
* start accepting proxy as an option ([#85][85])

[79]:https://github.com/airbrake/node-airbrake/pulls/79
[45]:https://github.com/airbrake/node-airbrake/pull/45
[51]:https://github.com/airbrake/node-airbrake/pull/51
[60]:https://github.com/airbrake/node-airbrake/pull/60
[61]:https://github.com/airbrake/node-airbrake/pull/61
[76]:https://github.com/airbrake/node-airbrake/pull/76
[78]:https://github.com/airbrake/node-airbrake/pull/78
[65]:https://github.com/airbrake/node-airbrake/pull/65
[68]:https://github.com/airbrake/node-airbrake/pull/68
[69]:https://github.com/airbrake/node-airbrake/pull/69
[74]:https://github.com/airbrake/node-airbrake/pull/74
[71]:https://github.com/airbrake/node-airbrake/pull/71
[72]:https://github.com/airbrake/node-airbrake/pull/72
[73]:https://github.com/airbrake/node-airbrake/pull/73
[84]:https://github.com/airbrake/node-airbrake/pull/84
[85]:https://github.com/airbrake/node-airbrake/pull/85
[86]:https://github.com/airbrake/node-airbrake/pull/86
[88]:https://github.com/airbrake/node-airbrake/pull/88
[89]:https://github.com/airbrake/node-airbrake/pull/89
[v0.4.0]: https://github.com/airbrake/node-airbrake/releases/tag/v0.4.0
[v0.4.1]: https://github.com/airbrake/node-airbrake/releases/tag/v0.4.1
[v1.0.0]: https://github.com/airbrake/node-airbrake/releases/tag/v1.0.0
[v1.0.1]: https://github.com/airbrake/node-airbrake/releases/tag/v1.0.1
[v1.0.2]: https://github.com/airbrake/node-airbrake/releases/tag/v1.0.2
[v1.0.3]: https://github.com/airbrake/node-airbrake/releases/tag/v1.0.3
[v1.1.0]: https://github.com/airbrake/node-airbrake/releases/tag/v1.1.0
[v1.2.0]: https://github.com/airbrake/node-airbrake/releases/tag/v1.2.0
[v1.2.1]: https://github.com/airbrake/node-airbrake/releases/tag/v1.2.1
[v1.2.2]: https://github.com/airbrake/node-airbrake/releases/tag/v1.2.2
[v1.3.0]: https://github.com/airbrake/node-airbrake/releases/tag/v1.3.0
[v2.0.0]: https://github.com/airbrake/node-airbrake/releases/tag/v2.0.0
[v2.0.1]: https://github.com/airbrake/node-airbrake/releases/tag/v2.0.1
[v2.1.0]: https://github.com/airbrake/node-airbrake/releases/tag/v2.1.0
[v2.1.1]: https://github.com/airbrake/node-airbrake/releases/tag/v2.1.1
[v2.1.2]: https://github.com/airbrake/node-airbrake/releases/tag/v2.1.2
[v2.1.3]: https://github.com/airbrake/node-airbrake/releases/tag/v2.1.3


test/run.js
#!/usr/bin/env node
var far = require('far').create();

far.add(__dirname);

var specificTest = process.argv[2];
if (typeof specificTest !== 'undefined' && specificTest !== 'all') {
  far.include(new RegExp('test-' + specificTest + '.js$'));
} else {
  far.include(/test-.*\.js$/);
}


var verbosity = process.argv[3];
if (typeof verbosity !== 'undefined') {
  verbosity = Number(verbosity);

  if (verbosity !== 1 && verbosity !== 2) {
    console.log('Please provide a verbosity value of 1 or 2');
    process.exit(1);
  }

  far.verbose(verbosity);
}
far.execute();


test/common.js
// Ignore rules for initialization file.
/* eslint global-require: 0 */
/* eslint import/no-unresolved: 0 */

var path = require('path');
var merge = require('lodash.merge');

// An account on the free plan specifically for testing this module.
exports.key = '81bbff95d52f8856c770bb39e827f3f6';
exports.projectId = '113743';

// Use custom config if available instead
try {
  merge(exports, require('./config'));
} catch (e) { console.log(e); }

exports.port = 8424;

var root = path.join(__dirname, '..');
exports.dir = {
  root: root,
  lib: root + '/lib'
};


test/test-host.js
var common = require('./common');
var airbrake = require(common.dir.root).createClient(common.projectId, common.key);
var assert = require('assert');
var os = require('os');

(function testDefaultHost() {
  assert.equal(airbrake.host, 'https://' + os.hostname());
}());

(function testPlainHost() {
  var err = new Error('oh no');
  var url = airbrake.contextJSON(err).url;
  assert.equal(url, airbrake.host.toLowerCase() + '/');
}());

(function testPartialErrUrl() {
  var err = new Error('oh no');
  err.url = '/foo';
  var url = airbrake.contextJSON(err).url;

  assert.equal(url, airbrake.host.toLowerCase() + err.url);
}());

(function testAbsoluteErrUrl() {
  var err = new Error('oh no');
  err.url = 'http://example.org/bar';
  var url = airbrake.contextJSON(err).url;

  assert.equal(url, err.url);
}());


lib/airbrake.js
var pkg = require('../package.json');

var fs = require('fs');
var os = require('os');
var EventEmitter = require('events').EventEmitter;
var request = require('request');
var stackTrace = require('stack-trace');
var merge = require('lodash.merge');
var execSync = require('child_process').execSync;
var url = require('url');

var truncator = require('../lib/truncator');


var HTTP_STATUS_CODES = require('http').STATUS_CODES;
var DEFAULT_SEVERITY = 'error';


function Airbrake() {
  this.key = null;
  this.projectId = null;

  this.host = 'https://' + os.hostname();
  this.env = process.env.NODE_ENV || 'development';
  this.whiteListKeys = [];
  this.blackListKeys = [];
  this.filters = [];
  this.projectRoot = process.cwd();
  this.appVersion = null;
  this.timeout = 30 * 1000;
  this.consoleLogError = false;

  this.proxy = null;
  this.protocol = 'https';
  this.serviceHost = process.env.AIRBRAKE_SERVER || 'api.airbrake.io';
  this.requestOptions = {};
  this.ignoredExceptions = [];
  this.exclude = [
    'type',
    'message',
    'arguments',
    'stack',
    'url',
    'session',
    'params',
    'component',
    'action',
    'ua'
  ];
}

merge(Airbrake.prototype, EventEmitter.prototype);

Airbrake.PACKAGE = (function() {
  var json = fs.readFileSync(__dirname + '/../package.json', 'utf8');
  return JSON.parse(json);
}());

Airbrake.createClient = function(projectId, key, env) {
  var instance = new this();
  instance.key = key;
  instance.env = env || instance.env;
  instance.projectId = projectId || instance.projectId;

  if (!instance.key || !instance.projectId) {
    throw new Error('Key or project ID missing during Airbrake.createClient()');
  }

  return instance;
};

Airbrake.prototype.expressHandler = function(disableUncaughtException) {
  var self = this;

  if (!disableUncaughtException) {
    process.on('uncaughtException', function(err) {
      self._onError(err, true);
    });
  }

  return function errorHandler(err, req, res, next) {
    var error = err;
    var requestObj = req;
    var responseObj = res;

    if (responseObj.statusCode < 400) responseObj.statusCode = 500;

    error.url = requestObj.url;
    error.action = requestObj.url;
    error.component = 'express';
    error.httpMethod = requestObj.method;
    error.params = requestObj.body;
    error.session = requestObj.session;
    error.ua = requestObj.get('User-Agent');

    self._onError(err, false);
    next(err);
  };
};

Airbrake.prototype.hapiHandler = function() {
  var self = this;
  var plugin = {
    register: function(server, options, next) {
      server.on('request-error', function(req, err) {
        var error = err;

        error.url =
          req.connection.info.protocol + '://' +
          req.info.host +
          req.url.path;

        error.action = req.url.path;
        error.component = 'hapi';
        error.httpMethod = req.method;
        error.params = request.params;
        error.ua = req.headers['user-agent'];

        self._onError(err, false);
      });

      next();
    }
  };

  plugin.register.attributes = { pkg: pkg };
  return plugin;
};

Airbrake.prototype._onError = function(err, die) {
  var self = this;
  var error = (err instanceof Error) ? err : new Error(err);
  self.log('Airbrake: Uncaught exception, sending notification for:');
  self.log(error.stack || error);

  self.notify(error, function(notifyErr, notifyUrl, devMode) {
    if (notifyErr) {
      self.log('Airbrake: Could not notify service.');
      self.log(notifyErr.stack);
    } else if (devMode) {
      self.log('Airbrake: Dev mode, did not send.');
    } else {
      self.log('Airbrake: Notified service: ' + notifyUrl);
    }

    if (die) {
      process.exit(1);
    }
  });
};

Airbrake.prototype.handleExceptions = function(die) {
  var self = this;
  var shouldDie = (typeof die === 'undefined') ? true : die;
  process.on('uncaughtException', function(err) {
    self._onError(err, shouldDie);
  });
};

Airbrake.prototype.handlePromiseRejections = function(die) {
  var self = this;
  var shouldDie = (typeof die === 'undefined') ? true : die;
  process.on('unhandledRejection', function(reason) {
    self._onError(reason, shouldDie);
  });
};

Airbrake.prototype.log = function(str) {
  if (this.consoleLogError) {
    console.error(str);
  }
};

Airbrake.prototype._sendRequest = function(body, cb) {
  var callback = this._callback(cb);

  var options = merge({
    method: 'POST',
    url: this.url('/api/v3/projects/' + this.projectId + '/notices?key=' + this.key),
    body: body,
    timeout: this.timeout,
    headers: {
      'Content-Length': body.length,
      'Content-Type': 'application/json',
      Accept: 'application/json'
    }
  }, this.requestOptions);

  request(options, function(requestErr, res, responseBody) {
    if (requestErr) {
      return callback(requestErr);
    }

    if (typeof responseBody === 'undefined') {
      return callback(new Error('invalid body'));
    }

    if (res.statusCode >= 300) {
      var status = HTTP_STATUS_CODES[res.statusCode];

      var explanation = responseBody.match(/<error>([^<]+)/i);
      explanation = (explanation)
        ? ': ' + explanation[1]
        : ': ' + responseBody;

      return callback(new Error(
        'Notification failed: ' + res.statusCode + ' ' + status + explanation
      ));
    }

    return callback(null, JSON.parse(responseBody).url);
  });
};

Airbrake.prototype.addFilter = function(filter) {
  this.filters.push(filter);
};

Airbrake.prototype.notify = function(err, cb) {
  var callback = this._callback(cb);
  var exit = false;

  this.ignoredExceptions.forEach(function(exception) {
    if (err instanceof exception) {
      exit = true;
    }
  });

  var notice = this.notifyJSON(err);

  this.filters.forEach(function(filter) {
    if (notice) {
      notice = filter(notice);
    }
  });

  if (exit || !notice) {
    return callback(null, null, false);
  }

  return this._sendRequest(truncator.jsonifyNotice(notice), callback);
};

Airbrake.prototype._callback = function(cb) {
  var self = this;
  return function(err) {
    if (cb) {
      cb.apply(self, arguments);
      return;
    }

    if (err) {
      self.emit('error', err);
    }
  };
};

Airbrake.prototype.url = function(path) {
  return this.protocol + '://' + this.serviceHost + path;
};

Airbrake.prototype.environmentJSON = function(err) {
  var cgiData = {};
  var self = this;

  if (this.whiteListKeys.length > 0) {
    Object.keys(process.env).forEach(function(key) {
      if (self.whiteListKeys.indexOf(key) > -1) {
        cgiData[key] = process.env[key];
      } else {
        cgiData[key] = '[FILTERED]';
      }
    });
  } else if (this.blackListKeys.length > 0) {
    Object.keys(process.env).forEach(function(key) {
      if (self.blackListKeys.indexOf(key) > -1) {
        cgiData[key] = '[FILTERED]';
      } else {
        cgiData[key] = process.env[key];
      }
    });
  }

  if (err.httpMethod) {
    cgiData.httpMethod = err.httpMethod;
  }

  Object.keys(err).forEach(function(key) {
    if (self.exclude.indexOf(key) >= 0) {
      return;
    }

    cgiData['err.' + key] = err[key];
  });

  cgiData['process.pid'] = process.pid;

  if (os.platform() !== 'win32') {
    // this two properties are *NIX only
    cgiData['process.uid'] = process.getuid();
    cgiData['process.gid'] = process.getgid();
  }

  cgiData['process.cwd'] = process.cwd();
  cgiData['process.execPath'] = process.execPath;
  cgiData['process.version'] = process.version;
  cgiData['process.argv'] = process.argv;
  cgiData['process.memoryUsage'] = process.memoryUsage();
  cgiData['os.loadavg'] = os.loadavg();
  cgiData['os.uptime'] = os.uptime();

  return cgiData;
};

Airbrake.prototype.contextJSON = function(err) {
  var context = {};
  context.notifier = {
    name: 'node-airbrake',
    version: Airbrake.PACKAGE.version,
    url: Airbrake.PACKAGE.homepage
  };

  context.environment = this.env;
  context.rootDirectory = this.projectRoot;
  context.os = os.type();
  context.hostname = os.hostname();
  context.url = url.resolve(this.host, err.url || '');
  context.userAgent = err.ua;
  context.component = err.component;
  context.action = err.action;
  context.severity = err.severity || DEFAULT_SEVERITY;

  return context;
};

Airbrake.prototype.notifyJSON = function(err) {
  var trace = stackTrace.parse(err);
  var self = this;

  return {
    errors: [
      {
        type: err.type || 'Error',
        message: err.message,
        backtrace: trace.map(function(callSite) {
          return {
            file: callSite.getFileName() || '',
            line: callSite.getLineNumber(),
            function: callSite.getFunctionName() || ''
          };
        })
      }],
    environment: self.environmentJSON(err),
    context: self.contextJSON(err),
    session: self.sessionVars(err),
    params: self.paramsVars(err)
  };
};

Airbrake.prototype.sessionVars = function(err) {
  return (typeof err.session === 'object')
    ? err.session
    : {};
};

Airbrake.prototype.paramsVars = function(err) {
  return (typeof err.params === 'object')
    ? err.params
    : {};
};

Airbrake.prototype.trackDeployment = function(params, cb) {
  var callback = cb;
  var deploymentParams = params || {};

  if (typeof deploymentParams === 'function') {
    callback = deploymentParams;
    deploymentParams = {};
  }

  deploymentParams = merge({
    key: this.key,
    env: this.env,
    user: process.env.USER,
    rev: execSync('git rev-parse HEAD').toString().trim(),
    repo: execSync('git config --get remote.origin.url').toString().trim()
  }, deploymentParams);

  var body = this.deploymentPostData(deploymentParams);

  var options = merge({
    method: 'POST',
    url: this.url('/api/v4/projects/' + this.projectId + '/deploys?key=' + this.key),
    body: body,
    timeout: this.timeout,
    headers: {
      'Content-Length': body.length,
      'Content-Type': 'application/json'
    },
    proxy: this.proxy
  }, this.requestOptions);

  var requestCallback = this._callback(callback);

  request(options, function(err, res, responseBody) {
    if (err) {
      return requestCallback(err);
    }

    if (res.statusCode >= 300) {
      var status = HTTP_STATUS_CODES[res.statusCode];
      return requestCallback(new Error(
        'Deployment failed: ' + res.statusCode + ' ' + status + ': ' + responseBody
      ));
    }

    return requestCallback(null, deploymentParams);
  });
};

Airbrake.prototype.deploymentPostData = function(params) {
  return JSON.stringify({
    version: 'v2.0',
    environment: params.env,
    username: params.user,
    revision: params.rev,
    repository: params.repo
  });
};

module.exports = Airbrake;
